sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs

```diff
@@ -0,0 +1,339 @@
+
+            var parts = SplitToCamelCase(identifierName).ToList();
+            for (int i = 0; i < parts.Count; i++)
+            {
+                string part = parts[i];
+                if (part.Length == 1 && part[0] == '_' && !areUnderscoresAllowed)
+                {
+                    continue;
+                }
+
+                var ideal = i == 0 ? HandleFirstTypePart(part, requireInitialI, allowInitialI, 1)
+                                    : SuggestLessUppercase(part, 1);
+
+                var acceptable = i == 0 ? HandleFirstTypePart(part, requireInitialI, allowInitialI, 2)
+                                         : SuggestLessUppercase(part, 2);
+
+                idealNameVariant.Append(ideal);
+                acceptableNameVariant.Append(acceptable);
+            }
+
+            suggestion = SuggestCapitalLetterAfterNonLetter(idealNameVariant);
+            var acceptableSuggestion = SuggestCapitalLetterAfterNonLetter(acceptableNameVariant);
+
+            return acceptableSuggestion == identifierName ||
+                   suggestion == identifierName;
+        }
+
+        private static string HandleFirstTypePart(string input,
+            bool requireInitialI, bool allowInitialI, int maxUppercase)
+        {
+            bool startsWithI = input[0] == 'I';
+
+            if (requireInitialI)
+            {
+                string prefix = startsWithI ? string.Empty : "I";
+                return prefix + SuggestLessUppercase(ToInitCap(input), maxUppercase);
+            }
+
+            string suggestionToProcess;
+            if (input.Length == 1 &&
+                !allowInitialI &&
+                startsWithI &&
+                IsCharUpper(input, 0) &&
+                !IsCharUpper(input, 1))
+            {
+                suggestionToProcess = ToInitCap(input.Substring(1));
+            }
+            else
+            {
+                suggestionToProcess = ToInitCap(input);
+            }
+
+            return SuggestLessUppercase(suggestionToProcess, maxUppercase);
+        }
+
+        private static string SuggestCapitalLetterAfterNonLetter(StringBuilder suggestion)
+        {
+            for (int i = 1; i < suggestion.Length; i++)
+            {
+                if (!char.IsLetter(suggestion[i - 1]) &&
+                    char.IsLower(suggestion[i]))
+                {
+                    suggestion[i] = char.ToUpperInvariant(suggestion[i]);
+                }
+            }
+            return suggestion.ToString();
+        }
+
+        private static string SuggestLessUppercase(string input, int maxUppercaseCount)
+        {
+            var upper = input.Take(maxUppercaseCount);
+            var lower = input.Skip(maxUppercaseCount).Select(char.ToLowerInvariant);
+
+            return new string(upper.Concat(lower).ToArray());
+        }
+
+        private static IEnumerable<string> SplitToCamelCase(string name)
```

Comment:
Not sure what the difference but we have an extension method with same (or close) name. I think it is worth adding a comment on the difference if we can't merge them.